src/ImageSharp/Formats/Jpeg/Components/Encoder/HuffmanScanEncoder.cs

```diff
@@ -262,43 +241,62 @@ public void Encode420<TPixel>(Image<TPixel> pixels, ref Block8x8F luminanceQuant
+        private void EncodeScanBaselineInterleaved<TPixel>(JpegFrame frame, SpectralConverter<TPixel> converter, CancellationToken cancellationToken)
             where TPixel : unmanaged, IPixel<TPixel>
         {
-            FastFloatingPointDCT.AdjustToFDCT(ref luminanceQuantTable);
-
-            this.huffmanTables = HuffmanLut.TheHuffmanLut;
-
-            // ReSharper disable once InconsistentNaming
-            int prevDCY = 0;
-
-            ImageFrame<TPixel> frame = pixels.Frames.RootFrame;
-            Buffer2D<TPixel> pixelBuffer = frame.PixelBuffer;
-            RowOctet<TPixel> currentRows = default;
-
-            var pixelConverter = new LuminanceForwardConverter<TPixel>(frame);
+            int mcu = 0;
+            int mcusPerColumn = frame.McusPerColumn;
+            int mcusPerLine = frame.McusPerLine;
 
-            for (int y = 0; y < pixels.Height; y += 8)
+            for (int j = 0; j < mcusPerColumn; j++)
             {
                 cancellationToken.ThrowIfCancellationRequested();
-                currentRows.Update(pixelBuffer, y);
 
-                for (int x = 0; x < pixels.Width; x += 8)
+                // Convert from pixels to spectral via given converter
+                converter.ConvertStrideBaseline();
+
+                // Encode spectral to binary
+                for (int i = 0; i < mcusPerLine; i++)
                 {
-                    pixelConverter.Convert(x, y, ref currentRows);
+                    // Scan an interleaved mcu... process components in order
+                    int mcuCol = mcu % mcusPerLine;
+                    for (int k = 0; k < frame.Components.Length; k++)
+                    {
+                        Component component = frame.Components[k];
+
+                        ref HuffmanLut dcHuffmanTable = ref this.dcHuffmanTables[component.DcTableId];
+                        ref HuffmanLut acHuffmanTable = ref this.acHuffmanTables[component.AcTableId];
 
-                    prevDCY = this.WriteBlock(
-                        QuantIndex.Luminance,
-                        prevDCY,
-                        ref pixelConverter.Y,
-                        ref luminanceQuantTable);
+                        int h = component.HorizontalSamplingFactor;
+                        int v = component.VerticalSamplingFactor;
 
+                        // Scan out an mcu's worth of this component; that's just determined
+                        // by the basic H and V specified for the component
+                        for (int y = 0; y < v; y++)
+                        {
+                            Span<Block8x8> blockSpan = component.SpectralBlocks.DangerousGetRowSpan(y);
+                            ref Block8x8 blockRef = ref MemoryMarshal.GetReference(blockSpan);
+
+                            for (int x = 0; x < h; x++)
+                            {
+                                int blockCol = (mcuCol * h) + x;
```

Comment:
JIT won't hoist `mcuCol * h` and this product is loop-invariant, so hoist it manually above the two for-loops.
I'd also make the type of `blockCol` `nint` for the `Unsafe.Add` below to avoid the sign-extending move.